doc/source/data/key-concepts.rst

```diff
@@ -0,0 +1,131 @@
+.. _data_key_concepts:
+
+Key Concepts
+============
+
+
+Datasets and blocks
+-------------------
+
+There are two main concepts in Ray Data:
+
+* Datasets
+* Blocks
+
+`Datasets` is the main user-facing Python API. They represent a distributed data collection and define data loading and processing operations. Users typically use the API by:
+
+1. Create a :class:`Dataset <ray.data.Dataset>` from external storage or in-memory data.
+2. Apply transformations to the data.
+3. Write the outputs to external storage or feed the outputs to training workers.
+
+The Dataset API is lazy, meaning that operations aren't executed until you call an action
+like :meth:`~ray.data.Dataset.show`. This allows Ray Data to optimize the execution plan
+and execute operations in parallel.
+
+*Blocks* are the basic unit of data that Ray Data operates on. A block is a contiguous
+subset of rows from a dataset.
+
+The following figure visualizes a dataset with three blocks, each holding 1000 rows.
+Ray Data holds the :class:`~ray.data.Dataset` on the process that triggers execution
+(which is usually the entrypoint of the program, referred to as the *driver*)
+and stores the blocks as objects in Ray's shared-memory
+:ref:`object store <objects-in-ray>`. Underneath the hood, blocks are represented as
+Pandas Dataframes or Arrow tables.
+
+.. image:: images/dataset-arch-with-blocks.svg
+..
+  https://docs.google.com/drawings/d/1kOYQqHdMrBp2XorDIn0u0G_MvFj-uSA4qm6xf9tsFLM/edit
+
+
+Plans and operators
+-------------------
+
+When a user writes a program using the Dataset API, a *logical plan* is constructed underneath the hood.
+
+A *logical plan* represents a sequence of data transformations, each of which is represented by a *logical operator*. For example, a ``Map`` operator represents applying a function to each row of the dataset, and a ``Project`` operator represents selecting a subset of columns from the dataset.
+
+Here is an example of a Ray Data program and its corresponding logical plan:
+
+.. code-block:: python
+
+    dataset = ray.data.read_csv("s3://my-bucket/my-file.csv")
+    dataset = dataset.map(lambda x: x + 1)
+    dataset = dataset.select_columns("col1")
+
+The logical plan for this program, which you can expect by calling ``print(dataset)``, is:
+
+.. code-block::
+
+    Project
+    +- Map(<lambda>)
+       +- Dataset(schema={...})
+
+When a dataset's execution plan is executed, the logical plan is optimized and transformed into a *physical plan* that in turn is also optimized. A *physical plan* is a graph of *physical operators*, which contain actual implementation of the data transformation and may also handle orchestration and execution across different Ray actors/workers. Read more about Ray actors and workers in :ref:`Ray Core Concepts <core-key-concepts>`.
```

Comment:
Let's capture the whole pipline:

1. Logical Optimization
2. Planning
3. Physical Optimization
4. Execution

And expand on these in details